packages/entity-server/src/routes/hierarchy/middleware/filter.ts

```diff
@@ -0,0 +1,63 @@
+/**
+ * Tupaia
+ * Copyright (c) 2017 - 2021 Beyond Essential Systems Pty Ltd
+ */
+
+import { Writable } from '../../../types';
+import { EntityFilter } from '../../../models';
+import { EntityFilterQuery, NestedFilterQueryFields } from '../types';
+
+const filterableFields: (keyof EntityFilterQuery)[] = [
+  'id',
+  'code',
+  'country_code',
+  'name',
+  'image_url',
+  'type',
+  'attributes_type',
+];
+const isFilterableField = (field: string): field is keyof EntityFilterQuery =>
+  (filterableFields as string[]).includes(field);
+
+const nestedFields: (keyof NestedFilterQueryFields)[] = ['attributes_type'];
+const isNestedField = (field: keyof EntityFilterQuery): field is keyof NestedFilterQueryFields =>
+  (nestedFields as (keyof EntityFilterQuery)[]).includes(field);
+
+type JsonBKey<T extends keyof NestedFilterQueryFields> = T extends `${infer Field}_${infer Key}`
+  ? `${Field}->>${Key}`
+  : T;
+
+const toJsonBKey = <T extends keyof NestedFilterQueryFields>(nestedField: T): JsonBKey<T> => {
+  const [field, value] = nestedField.split('_');
+  return [field, value].join('->>') as JsonBKey<T>;
```

Comment:
Have to cast explicitly here unfortunately as `string.split()` returns a generic `string` type. It's only a matter of time before I write a custom `split()` function that can return the correctly typed array of a string literal type 😝